### PR TITLE
License Export

### PR DESCRIPTION
**Pull request by henrikt-ma**
This is MCP-0029 _License Export_. Currently in _Draft_ state to reflect state of being under development.

This PR is the continuation of the issue #2217.

### CODE REVIEW

**Comment by HansOlsson** [comment]
In order to avoid confusion at least the summary should make it clear that it's not related to licensing (and other proposals for IP-protection) but just about including the license texts.

**Comment by HansOlsson** [comment]
Several of the links in the document seem to be broken.

**Comment by beutlich** [comment]
> Several of the links in the document seem to be broken.

Resolved by 70f0c49fa2a10bd681811ce626e8b965bdffd239.

File: RationaleMCP/0029/ReadMe.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+### Experience with Prototype
+
+On the library side, the necessary changes have been applied to library _ExternData_.
+The following commits at the Git repository at https://github.com/modelica-3rdparty/ExternData demonstrate the necessary steps to update a Modelica libarry.
+
+1. https://github.com/modelica-3rdparty/ExternData/commit/7522046dce6d20fe01a85a559b02fd9d8878bdd2
+   - All license files were added as text files to a new directory ExternData/Resources/Licenses.
+     The file format is not restricted to Text with a special encoding, but any file format (e.g., Mark-Down, PDF, HTML) basically can be supported.
+     The license file LICENSE_ExternData.txt is the corresponding license of the copyright holder of the Modelica package, all other license files reflect third-party software components.
+   - It is recommended to use ModelicaPackage/Resource/Licenses as default directory for the license files.
```
Comment (otronarp) [inline RationaleMCP/0029/ReadMe.md:53]: What does _default_ mean here, is this a recommendation to users to put their files here or is the tool expected to do anything special with this _default_ directory?

Comment (beutlich) [reply RationaleMCP/0029/ReadMe.md:53]: Either use this directory or the one specified by the License annotation.

Comment (otronarp) [reply RationaleMCP/0029/ReadMe.md:53]: It's still no very clear what you mean, but my interpretation is that you should remove the word _default_ from that sentence.

Comment (HansOlsson) [reply RationaleMCP/0029/ReadMe.md:53]: Agreed. The recommended makes it clear that it can be overriden.

Comment (HansOlsson) [inline RationaleMCP/0029/ReadMe.md:53]:
```suggestion
   - It is recommended to use ModelicaPackage/Resource/Licenses as the directory for the license files.
```
As proposed by Otto.

**Comment by HansOlsson** [comment]
Design meeting:

Seems good - try to make ready for next release.

**Comment by HansOlsson** [comment]
Since we now have a prototype implementation I would suggest:

- [ ] Merge master into this branch (to avoid merge conflicts later).
- [ ] Add the few lines needed for the specification.
- [ ] Then review it.

**Comment by HansOlsson** [comment]
> Since we now have a prototype implementation I would suggest:
> 
> * [ ]  Merge master into this branch (to avoid merge conflicts later).
> * [ ]  Add the few lines needed for the specification.
> * [ ]  Then review it.

I think it would be good if we could progress on this simple (and fairly uncontroversial) MCP.

**Comment by beutlich** [comment]
Please note, that I rebased and force-pushed the PR branch.

File: chapters/functions.tex [hunk 2]
```diff
@@ -2398,6 +2398,11 @@ \subsection{Annotations for External Functions}\label{annotations-for-external-l
   {\lstinline!annotation(SourceDirectory="modelica:/ModelicaLibraryName/Resources/Source")!}\annotationindex{SourceDirectory}, gives the location for source files.
   The preceding one is the default and need not be specified; but another location could be specified by using an URI name for the source directory, see \cref{external-resources}.
   It is not specified how they are built.
+\item
+  The {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives the license files for the function.
+  The license files are human-readable files containing the license conditions.
```
Comment (henrikt-ma) [inline chapters/functions.tex:2403]: The annotation syntax only allows specifying a single file, which should be matched by the text:
```suggestion
  The {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license file for the function.
  License files are human-readable files containing the license conditions.
```

Comment (henrikt-ma) [reply chapters/functions.tex:2403]: This begs the question: Can a function have multiple `License`-annotations?

File: chapters/annotations.tex [hunk 3]
```diff
@@ -2324,6 +2324,15 @@ \subsection{Licensing}\label{licensing}
 \end{example}
 
 
+\section{License Texts}\label{license-texts}
+
+For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives the license files for the class.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2329]: Match the singular example by singular in the surrounding text:
```suggestion
For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license file for the class.
```

File: chapters/annotations.tex [hunk 4]
```diff
@@ -2324,6 +2324,15 @@ \subsection{Licensing}\label{licensing}
...
+The annotation may give a scalar value or an array of values.
+The license files are human-readable files containing the license conditions.
+External functions may also contain this annotation (with similar contents), see \cref{annotations-for-external-libraries-and-include-files}.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2332]: Better with _content_?
```suggestion
External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.
```

File: chapters/functions.tex (hunk 2 above)
Comment (henrikt-ma) [reply chapters/functions.tex:2403]: I'll answer my own question by noting that it would probably be better to just refer to how the annotation's definition for a top-level class than to partially repeat the definition here:
```suggestion
  The {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license files for the function.
  It is analogous to the \lstinline!License!-annotation for a top-level class, see \cref{license-texts}.
```

File: chapters/functions.tex [hunk 5]
```diff
@@ -2398,6 +2398,11 @@ \subsection{Annotations for External Functions}\label{annotations-for-external-l
...
+  Top-level classes may also contain this annotation (with similar contents), see \cref{license-texts}.
+  When generating a distributable version for a translated model the license files for used functions and top-level classes should be included.
```
Comment (henrikt-ma) [inline chapters/functions.tex:2405]: Rely on `\cref{license-texts}` and avoid repetition:
```suggestion
```

File: chapters/annotations.tex (hunk 4 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2332]: Yes, if we consider the meaning of the annotation - not the letters.

File: chapters/functions.tex (hunk 2 above)
Comment (HansOlsson) [reply chapters/functions.tex:2403]: The last line is still useful.

File: chapters/functions.tex (hunk 5 above)
Comment (HansOlsson) [reply chapters/functions.tex]: This has been removed and the proposed line above added.

**Comment by HansOlsson** [comment]
To me this seems ready, and we can move forward with this and start formal review soonish.
We have a prototype in Dymola and a library using it already.

File: chapters/functions.tex [hunk 6]
```diff
@@ -2398,6 +2398,10 @@ \subsection{Annotations for External Functions}\label{annotations-for-external-l
   {\lstinline!annotation(SourceDirectory="modelica:/ModelicaLibraryName/Resources/Source")!}\annotationindex{SourceDirectory}, gives the location for source files.
   The preceding one is the default and need not be specified; but another location could be specified by using an URI name for the source directory, see \cref{external-resources}.
   It is not specified how they are built.
+\item
+  The {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives the license file for the function.
+  The license files are human-readable files containing the license conditions.
```
Comment (henrikt-ma) [inline chapters/functions.tex:2403]: Rely even more on `\cref{license-texts}`:
```suggestion
```

File: chapters/functions.tex [hunk 7]
```diff
@@ -2398,6 +2398,10 @@ \subsection{Annotations for External Functions}\label{annotations-for-external-l
...
+  It is analogous to the \lstinline!License!-annotation for a top-level class, see \cref{license-texts}.
```
Comment (henrikt-ma) [inline chapters/functions.tex:2404]: I realized earlier today that annotation-names are not hyphened:
```suggestion
  It is analogous to the \lstinline!License! annotation for a top-level class, see \cref{license-texts}.
```

File: chapters/annotations.tex [hunk 8]
```diff
@@ -2324,6 +2324,16 @@ \subsection{Licensing}\label{licensing}
...
 
 
+\section{License Texts}\label{license-texts}
+
+For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license file for the class.
+The annotation may give a scalar value or an array of values.
+The license files are human-readable files containing the license conditions.
+If the annotation specifies an array of license files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).
+External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.
+When generating a distributable version for a translated model the license files for used functions and top-level classes should be included.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2334]: Is it a conscious decision to _not_ require that tools include the license files?

Comment (HansOlsson) [reply chapters/annotations.tex:2334]: There are several possible reasons:
1. The vagueness of "distributable version" makes it odd to have a hard requirement. (Is an executable distributable if it can be copied or does it require more? Is it distributed if distributed internally in a company?)
2. A tool may have options to handle the license files in other ways; if you generate a distributable version for embedding in a device the license files should be delivered in a completely different way: verified by a legal department, and some of the license files might be skipped (after review).
3. If the library is internal to the ones translating the model it makes sense that they license owners can skip their own licenses.

In all cases one could just remove the license files afterwards, so people can anyway circumvent it.

Comment (HansOlsson) [reply chapters/annotations.tex:2334]: Note that 'should' doesn't mean that it can be skipped at will, just that there _may_ exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

File: chapters/annotations.tex [hunk 9]
```diff
@@ -2324,6 +2324,16 @@ \subsection{Licensing}\label{licensing}
 \end{example}
 
 
+\section{License Texts}\label{license-texts}
```
Comment (beutlich) [inline chapters/annotations.tex:2327]: What do you think, could we also add a better example here (similar as in 18.9.2)?

Comment (HansOlsson) [reply chapters/annotations.tex:2327]: We could, but I think it can wait.

**Comment by HansOlsson** [comment]
Can we move forward with this MCP? ("License export")

File: chapters/annotations.tex (hunk 8 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:2334]: A nice way to resolve this would be to include a non-normative explanation of why a tool might choose to not include the license files.

File: RationaleMCP/0029/ReadMe.md (hunk 1 above)
Comment (HansOlsson) [reply RationaleMCP/0029/ReadMe.md]: File has been updated.

File: chapters/functions.tex (hunk 2 above)
Comment (HansOlsson) [reply chapters/functions.tex]: Seems resolved.

**Comment by HansOlsson** [comment]
There are some things that could be made even clearer, but I think this is ready to be evaluated.

Use emojis (face left-below) to vote (closes day before next phone meeting):
1.  Thumbs up: Ok to move forward to "Under evaluation". If so please provide a Review as well (we need three to move forward).
2. Thumbs down: Don't want this at all.
3.  Eyes: Need more discussion.

File: chapters/annotations.tex (hunk 8 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2334]: -> @HansOlsson write explanation.

**Comment by HansOlsson** [comment]
Reviewers: Henrik, Elena

**Review by beutlich — COMMENTED** [review]
After dealing with OSS licenses for some years, I prefer to have the utilized third-party licenses explicit set. See https://github.com/modelica/ModelicaStandardLibrary/pull/4268 for what I mean. This way it is clear which licenses are directly produced by the original authors and which one are pulled-in by third-party OSS.

**Comment by henrikt-ma** [comment]
> After dealing with OSS licenses for some years, I prefer to have the utilized third-party licenses explicit set. See [modelica/ModelicaStandardLibrary#4268](https://github.com/modelica/ModelicaStandardLibrary/pull/4268) for what I mean. This way it is clear which licenses are directly produced by the original authors and which one are pulled-in by third-party OSS.

It doesn't sound like a bad idea to me, but I think we need a concrete proposal here, and I don't think https://github.com/modelica/ModelicaStandardLibrary/pull/4268 gives all the answers. In my opinion, having a _Licenses/Third-party_ suggests that no license files should be placed directly under _Licenses_, but that there should also be a directory for non-third-party licenses next to _Third-party_. This would make it clear that one always needs to make an active choice of where to place the file, and avoiding that the use of _Third-party_ somehow ends up being interpreted as voluntary.

**Comment by HansOlsson** [comment]
> Reviewers: Henrik, Elena

Would be good if we could those reviews to move this forward.

**Comment by henrikt-ma** [comment]
> Even if I think that it is a good idea, I don't see that we need to standardize this, but that library authors can handle that themselves - including the naming of the license files, and whether non-3rd party licenses should be in a separate directory, or directly under licenses.

I agree that the standard should not prescribe fixed locations for the license files, as it goes against the idea of using a Modelica URI to locate the file.  (If there were fixed locations, we should have simply stated that all relevant license files (and only those) should be placed in `modelica:/ModelicaLibraryName/Resources/Licenses`.)

If we would like to have special recognition of license files coming from third-party dependencies in the future, we could add another annotation: `annotation(ThirdPartyLicense="modelica:/ModelicaLibraryName/Resources/ThirdPartyThing/TheirLicense.txt")`

File: chapters/annotations.tex (hunk 8 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:2334]: @HansOlsson ping

**Comment by beutlich** [comment]
I agree with your comments @HansOlsson and @henrikt-ma. The main intention of https://github.com/modelica/ModelicaStandardLibrary/pull/4268 was to increase visibility and transparency what's MA and what's third-party as "explicit is better than implicit".

File: chapters/annotations.tex (hunk 8 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2334]: Have done

File: chapters/annotations.tex [hunk 10]
```diff
@@ -2324,6 +2324,18 @@ \subsection{Licensing}\label{licensing}
...
+
+For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license file for the class.
+The annotation may give a scalar value or an array of values.
+The license files are human-readable files containing the license conditions.
+If the annotation specifies an array of license files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).
+External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.
+When generating a distributable version for a translated model the license files for used functions and top-level classes should be included.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2334]: New paragraph to make exception and the final sentence more clearly referring to the _should_:
```suggestion

When generating a distributable version for a translated model the license files for used functions and top-level classes should be included.
```

File: chapters/annotations.tex [hunk 11]
```diff
@@ -2324,6 +2324,18 @@ \subsection{Licensing}\label{licensing}
...
+Exceptions could be that licenses will be distributed in another way, either the same ones -- or there is an agreement to use different licenses (e.g., for internal libraries).
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2335]: I find this hard to parse and would prefer:
```suggestion
Exceptions could be that licenses will be distributed in another way, or that there is an agreement to include different licenses (e.g., for internal libraries).
```

File: chapters/annotations.tex [hunk 12]
```diff
@@ -2324,6 +2324,18 @@ \subsection{Licensing}\label{licensing}
...
+Note that the collection of license files is only intended as a help, thus there is no need for a strict definition of distributable version or stricter rules for including the license files.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2336]: I'm trying to see for whom this help is… Am I getting it wrong if I believe this is what is meant?
```suggestion
Note that the \lstinline!License!-annotation only allows the library developer to assist the tool in doing what is legally required; in the end it is a user responsibility to ensure that required license files accompany distributable versions of models.
```

I'd like to skip the part about _strict definition_, since I personally would find it helpful if the specification made it a tool responsibility in at least some of the most common situations, so that I would have the specification on my side when complaining to my tool vendor until they implement this.

File: chapters/annotations.tex (hunk 8 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:2334]: OK, let's continue at the added content.  Resolving.

**Comment by henrikt-ma** [comment]
The structure of the `License`-annotation is identical with the `styleSheets` which was recently added in #3409.  Shouldn't we be introducing `License` in the same modern style (where the inclusion of `constant` would depend on whether #3448 is merged first)?
```
constant String[:] Licenses "URIs pointing to license files";
```

File: chapters/annotations.tex (hunk 10 above)
Comment (eshmoylova) [inline chapters/annotations.tex:2334]: Should "distributable translated model" be clarified even a little bit? Adding "e.g., an FMU" at least? We have read the ReadMe.md and know where it's coming from, but I am not sure it would be clear as it stands in the specification. Translation that is mentioned in the specification is explained as "expansion of inherited base classes; parameterization of base classes,..." (section 1.2).

File: chapters/annotations.tex [hunk 13]
```diff
@@ -2324,6 +2324,18 @@ \subsection{Licensing}\label{licensing}
 \end{example}
 
 
+\section{License Texts}\label{license-texts}
+
+For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license file for the class.
+The annotation may give a scalar value or an array of values.
+The license files are human-readable files containing the license conditions.
+If the annotation specifies an array of license files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).
```
[the remark this answers is not in the document]
Comment (HansOlsson) [reply chapters/annotations.tex:2332]: The person(s) distributing the derivative work depending on them I would say.

When making a library with license files the one building/distributing that library should (of course) ensure that the licenses are valid for the library.

File: chapters/annotations.tex (hunk 10 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2334]: Makes sense.

**Comment by HansOlsson** [comment]
Language group: "license text files" seems best, but think of other options during the week. Alternatives are "license conditions", "license agreements", and possibly more.

**Comment by HansOlsson** [comment]
> Language group: "license text files" seems best, but think of other options during the week. Alternatives are "license conditions", "license agreements", and possibly more.

Looking more I think that even if "license agreements" could work it often refers to other agreements (for patents, trademarks, etc), but these licenses are primarily a copyright-license, so if we wanted to be clear it should be "copyright license agreements".

However it also seems to depend on the kind of license: https://opensource.org/licenses - just calls it licenses, whereas commercial licenses are often called license agreements (including having you actually sign that you agree to them).

**Comment by beutlich** [comment]
I'd simply call it "license file". See also https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository#determining-the-location-of-your-license

**Comment by henrikt-ma** [comment]
> I'd simply call it "license file". See also https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository#determining-the-location-of-your-license

We're looking for something which is not called just _license file_ to avoid collision with the `licenseFile`-annotation.  I've updated it now to _license text file_ according to https://github.com/modelica/ModelicaSpecification/pull/2900#issuecomment-1954375478, but I'd be in favor of something more clearly distinct from _license file_.

**Comment by beutlich** [comment]
> > I'd simply call it "license file". See also https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository#determining-the-location-of-your-license
> 
> We're looking for something which is not called just _license file_ to avoid collision with the `licenseFile`-annotation. I've updated it now to _license text file_ according to [#2900 (comment)](https://github.com/modelica/ModelicaSpecification/pull/2900#issuecomment-1954375478), but I'd be in favor of something more clearly distinct from _license file_.

Hm, is a file License.md also a _license text file_ or is it a _license markdown file_?

**Comment by HansOlsson** [comment]
> > > I'd simply call it "license file". See also https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository#determining-the-location-of-your-license
> > 
> > 
> > We're looking for something which is not called just _license file_ to avoid collision with the `licenseFile`-annotation. I've updated it now to _license text file_ according to [#2900 (comment)](https://github.com/modelica/ModelicaSpecification/pull/2900#issuecomment-1954375478), but I'd be in favor of something more clearly distinct from _license file_.
> 
> Hm, is a file License.md also a _license text file_ or is it a _license markdown file_? I'd even claim that the annotation name has to match the other wording.

I would say that "license text file" refers to the logical contents - i.e., that the "file" contains the "license text" in these cases.
So, whether the file _format_ is plain text, markdown, pdf, or a jpeg with a photo of the license text(!) shouldn't matter here.

It may still be that we should use another name.

**Comment by beutlich** [comment]
AI help for the rescue:

grafik

**Comment by HansOlsson** [comment]
Another solution would be to describe the other "license files" as something else (the files that actually provide computer-readable licenses).

However, looking closely it seems that when using FlexNet and similar programs the computer-readable files for licenses are often called "license files".

File: chapters/annotations.tex [hunk 14]
```diff
@@ -2324,6 +2324,19 @@ \subsection{Licensing}\label{licensing}
...
 
+\section{License Texts}\label{license-texts}
+
+For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license text file for the class.
+The annotation may give a scalar value or an array of values.
+The license text files are human-readable files containing the license conditions.
+If the annotation specifies an array of license text files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).
+External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.
+
+When generating a distributable version for a translated model the license text files for used functions and top-level classes should be included.
```
Comment (HansOlsson) [inline chapters/annotations.tex:2335]:
```suggestion
When generating a distributable version (e.g., an FMU) for a translated model the license text files for used functions and top-level classes should be included.
```
As suggested.

File: chapters/annotations.tex [hunk 15]
```diff
@@ -2324,6 +2324,19 @@ \subsection{Licensing}\label{licensing}
...
+Exceptions could be that licenses will be distributed in another way, or that there is an agreement to include different licenses (e.g., for internal libraries).
+Note that the collection of license text files is only intended as a help, thus there is no need for a strict definition of distributable version or stricter rules for including the license text files.
```
Comment (HansOlsson) [inline chapters/annotations.tex:2337]:
```suggestion
Note that the \lstinline!License!-annotation only allows the library developer and tool to assist the user in doing what is legally required; in the end it is a user responsibility to ensure that required license files accompany distributable versions of models.
```
As suggested.

File: chapters/annotations.tex (hunk 13 above)
[the remark this answers is not in the document]
Comment (HansOlsson) [reply chapters/annotations.tex]: I would say the general statement about this only being an aid for the user covers this as well.

File: chapters/annotations.tex (hunk 14 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2335]: @eshmoylova - your comment above.

File: chapters/annotations.tex (hunk 15 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2337]: @henrikt-ma your comment above with changes.

File: chapters/annotations.tex (hunk 12 above)
Comment (HansOlsson) [reply chapters/annotations.tex]: Turned into suggestion below.

File: chapters/annotations.tex (hunk 14 above)
Comment (eshmoylova) [reply chapters/annotations.tex:2335]: That would work for me.

Comment (beutlich) [reply chapters/annotations.tex:2335]: Is FMU already declared/defined in the MoLang Spec?

Comment (HansOlsson) [reply chapters/annotations.tex:2335]: @eshmoylova  still ok?

File: chapters/annotations.tex (hunk 15 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2337]: @henrikt-ma are you ok with change based on your comment?

File: chapters/annotations.tex (hunk 14 above)
Comment (eshmoylova) [reply chapters/annotations.tex:2335]: FMU is not defined or used so far. But we refer to C89 and C99, for example, without providing any explanation or links. So, it should be acceptable to refer to FMU like that as well. It might be nicer to make it a link.

Comment (beutlich) [reply chapters/annotations.tex:2335]: Yes, that was the intention of my comment: FMU should be declared and not considered as given.

File: chapters/annotations.tex (hunk 15 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:2337]: Yes, it is in the same spirit.

File: chapters/annotations.tex (hunk 14 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:2335]: If linked, I think it should be handled similar to gettext; as a citation with the link given in the bibliography.

Comment (HansOlsson) [reply chapters/annotations.tex:2335]: We can add the link later, if needed, but is the current proposed text good enough for the moment?

Comment (eshmoylova) [reply chapters/annotations.tex:2335]: Looks good.

**Review by eshmoylova — APPROVED** [review]

**Comment by beutlich** [comment]
In preparation of the voting...

* Is the term `license text files` finally resolved/decided? It rather looks still pending to me. If it is decided it is inconsistent in one location presumablely and inconsistent with FMI (where it is called `license files` in https://fmi-standard.org/docs/3.0.1/#structure-of-zip with example given). Note, that I am alo in favour to name it `license files` and give an example.
* Apart from this naming issue, it feels confirmative that my MCP-0029 is on the finish line. As usual, the rationale was written within some days, but the final text takes years...

**Comment by otronarp** [comment]
In the discussion above it sounds like the location is assumed to be given by a  modelica URI. If that's the case, perhaps that should be spelled out more in text and not only in the code snippet. Like it is for the other external function annotations.

> ...location could be specified by using an URI name for the library directory, see section 13.5.

**Comment by HansOlsson** [comment]
> In preparation of the voting...
> 
> * Is the term `license text files` finally resolved/decided? 

Yes. It was discussed at the latest phone meeting, and nothing better was found.

> * Apart from this naming issue, it feels confirmative that my MCP-0029 is on the finish line. As usual, the rationale was written within some days, but the final text takes years...

Yes, that is a constant problem. I think this one was unnecessarily delayed since "license" evoked a lot of other license-issues.

**Comment by henrikt-ma** [comment]
> In the discussion above it sounds like the location is assumed to be given by a modelica URI. If that's the case, perhaps that should be spelled out more in text and not only in the code snippet. Like it is for the other external function annotations.
> 
> > ...location could be specified by using an URI name for the library directory, see section 13.5.

This is a good suggestion.  Can we consider it _Typos and similar trivial changes_ so that we can fix it during the evaluation period?

File: chapters/annotations.tex [hunk 16]
```diff
@@ -2324,6 +2324,19 @@ \subsection{Licensing}\label{licensing}
 \end{example}
 
 
+\section{License Texts}\label{license-texts}
+
+For a top-level class the {\lstinline!annotation(License="modelica:/ModelicaLibraryName/Resources/Licenses/MyLicense.txt")!}\annotationindex{License}, gives a license text file for the class.
+The annotation may give a scalar value or an array of values.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2330]: Here is a concrete proposal for addressing @otronarp's https://github.com/modelica/ModelicaSpecification/pull/2900#issuecomment-2022343060:
```suggestion
The annotation may give a scalar string or an array of strings, where each string is a URI referencing a license text file, see \cref{external-resources}.
```

Comment (HansOlsson) [reply chapters/annotations.tex:2330]: As I understand it this is the only minor issue to discuss after it was accepted.
We might either accept the PR without it and then have a separate PR for this proposal, or amend the PR (arguing that it is just a trivial change).

Comment (beutlich) [reply chapters/annotations.tex:2330]: I also raised one other concern previously:

> it is inconsistent in one location presumablely

I belive it can be easily addressed now.

File: chapters/annotations.tex [hunk 17]
```diff
@@ -2324,6 +2324,19 @@ \subsection{Licensing}\label{licensing}
...
+The license text files are human-readable files containing the license conditions.
+If the annotation specifies an array of license text files the conditions in all of them must be satisfied (i.e., it is not intended for dual-licensing).
+External functions may also contain this annotation (with similar content), see \cref{annotations-for-external-libraries-and-include-files}.
+
+When generating a distributable version for a translated model the license text files for used functions and top-level classes should be included.
+Exceptions could be that licenses will be distributed in another way, or that there is an agreement to include different licenses (e.g., for internal libraries).
+Note that the \lstinline!License!-annotation only allows the library developer and tool to assist the user in doing what is legally required; in the end it is a user responsibility to ensure that required license files accompany distributable versions of models.
```
Comment (HansOlsson) [inline chapters/annotations.tex:2337]:
```suggestion
Note that the \lstinline!License!-annotation only allows the library developer and tool to assist the user in doing what is legally required; in the end it is a user responsibility to ensure that required license text files accompany distributable versions of models.
```
